+--------------------------------------------------------------------------------------------------------------+
| Note: Google Genomics is now Cloud Life Sciences.                                                            |       
| The Google Genomics Cookbook on Read the Docs is not actively                                                |
| maintained and may contain incorrect or outdated information.                                                |
| The cookbook is only available for historical reference. For                                                 |
| the most up to date documentation, view the official Cloud                                                   |
| Life Sciences documentation atÂ https://cloud.google.com/life-sciences.                                       |
|                                                                                                              |
| Also note that much of the Genomics v1 API surface has been                                                  |
| superseded by `Variant Transforms <https://cloud.google.com/life-sciences/docs/how-tos/variant-transforms>`_ |
| and `htsget <https://cloud.google.com/life-sciences/docs/how-tos/reading-data-htsget>`_.                     |
+--------------------------------------------------------------------------------------------------------------+

BioC 2015: Where Software and Biology Connect
=============================================

.. comment: begin: goto-read-the-docs

.. container:: visible-only-on-github

   +-----------------------------------------------------------------------------------+
   | **The properly rendered version of this document can be found at Read The Docs.** |
   |                                                                                   |
   | **If you are reading this on github, you should instead click** `here`__.         |
   +-----------------------------------------------------------------------------------+

.. _RenderedVersion: http://googlegenomics.readthedocs.org/en/latest/workshops/bioc-2015.html

__ RenderedVersion_

.. comment: end: goto-read-the-docs

*This workshop was presented at the annual* `Bioconductor Developer's Conference <http://www.bioconductor.org/help/course-materials/2015/BioC2015/>`_.

Google has some pretty amazing big data computational "hammers" that they have been applying to search and video data for a long time.  In this workshop we take those same hammers and apply them to whole genome sequences.

We will work with both the :doc:`/use_cases/discover_public_data/1000_genomes` reads and variants and also the :doc:`/use_cases/discover_public_data/platinum_genomes` gVCF variants.

We do this all from the comfort of the R prompt using common packages including `VariantAnnotation`_, `ggbio`_, `ggplot2`_, `dplyr`_, `bigrquery`_, and the new Bioconductor package `GoogleGenomics`_ which provides an R interface to Google's implementation of the `Global Alliance for Genomics and Health API`_.

And we'll do this in a reproducible fashion running RMarkdown files via `Dockerized Bioconductor`_ running on `Google Compute Engine`_ VMs!

.. contents::

Get Started with Google Cloud Platform
--------------------------------------

Create a Google Cloud Platform project
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. include:: /includes/gcp_signup.rst

Enable APIs
^^^^^^^^^^^

Enable all the Google Cloud Platform APIs we will use in this workshop by clicking on this `link <https://console.cloud.google.com/flows/enableapi?apiid=genomics,bigquery>`_.

Install gcloud
^^^^^^^^^^^^^^

.. include:: /includes/gcloud_setup.rst

Set up Bioconductor
-------------------------------------------------------------

.. include:: /includes/bioconductor_deployment_sidebar.rst

To further the goals of `reproducibility, ease of use, and convenience <http://www.bioconductor.org/help/docker/>`_, you can run this codelab in a Bioconductor Docker container deployed to `Google Compute Engine`_.  But this codelab `can be run from anywhere <https://github.com/googlegenomics/bioconductor-workshop-r>`_ since all the heavy lifting is happening in the cloud regardless of where R is running.

Bioconductor maintains Docker containers with R, Bioconductor packages, and RStudio Server all ready to go!  Its a great way to set up your R environment quickly and start working.  The instructions are below but if you want to learn more, see http://www.bioconductor.org/help/docker/.

1. Click on `click-to-deploy Bioconductor`_ to navigate to the deployer page on the Cloud Platform Console.
2. In field *Docker Image* choose item ``custom``.
3. Click on *More* to display the additional form fields.
4. In field *Custom docker image* paste in value ``gcr.io/bioc_2015/devel_sequencing``.
5. Click on the *Deploy Bioconductor* button.
6. Follow the post-deployment instructions to log into RStudioServer via your browser!

.. container:: toggle

    .. container:: header

        If you prefer to run this docker container locally, click here to **Show/Hide Instructions**

    .. container:: content

      .. include:: /includes/bioconductor_docker_details.rst

.. container:: toggle

    .. container:: header

        If you prefer to setup R manually instead, click here to **Show/Hide Instructions**

    .. container:: content

      .. include:: /includes/bioconductor_workshop_r_setup.rst

Run the Codelabs
------------------------------------------------------

(1) View the workshop documentation.

.. code::

  help(package="GoogleGenomicsBioc2015Workshop")

(2) Click on *"User guides, package vignettes and other documentation."*

(3) Early on in the workshop you will need an API_KEY.  You can get this by clicking on this link: https://console.cloud.google.com/project/_/apiui/credential

(4) Click on vignette *"Bioc2015Workshop"* and follow the instructions there to run the vignettes **line-by-line** or **chunk-by-chunk**!

  * To run **line-by-line**, put your cursor on the desired line  and click the "Run" button or use keyboard shortcuts for Windows/Linux: ``Ctrl+Enter`` and Mac: ``Command+Enter``.

  * To run **chunk-by-chunk**, put your cursor in the desired chunk and click the "Chunks -> Run Current Chuck" button. or use keyboard shortcuts for Windows/Linux: ``Ctrl+Alt+C`` and Mac: ``Command+Option+C``.

  .. image:: /_static/run_rmarkdown.png
    :alt: Run Rmarkdown

If you just want to read the rendered results of the four codelabs, here they are:

* `Working with Reads <http://www.bioconductor.org/packages/devel/bioc/vignettes/GoogleGenomics/inst/doc/PlottingAlignments.html>`_
* `Working with Variants <http://www.bioconductor.org/packages/devel/bioc/vignettes/GoogleGenomics/inst/doc/AnnotatingVariants.html>`_
* `Analyzing Variants with BigQuery <https://github.com/googlegenomics/bioconductor-workshop-r/blob/master/inst/doc/BigQueryDemo.md>`_
* `Data Analysis using Google Genomics <https://github.com/googlegenomics/bioconductor-workshop-r/blob/master/inst/doc/AllModalitiesDemo.md>`__, also available on YouTube:

  * brief video `Google Genomics: Data Analysis Overview <https://www.youtube.com/watch?v=vINpqxhcTt0>`_
  * extended video `Google Genomics Codelab: Data Analysis in R <https://www.youtube.com/watch?v=tPH5PwjzhBM>`_

Looking for more to try?
------------------------
* Try these samples on different datasets :doc:`/use_cases/discover_public_data/index`.
* Find more example BigQuery queries in:

  * `getting-started-bigquery`_
  * `bigquery-examples`_
  * `Quality Control using Google Genomics`_

* Run a `Google Cloud Dataflow`_ pipeline:

  * :doc:`/use_cases/compute_identity_by_state/index`
  * :doc:`/use_cases/compute_principal_coordinate_analysis/1-way-pca`

Wrap up
-------------

"Stop" or "Delete" your VM
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. include:: /includes/c2d_deployment_teardown.rst

Stay Involved
^^^^^^^^^^^^^

* Collaborate on the `GA4GH`_ specification.
* :doc:`/mailinglist`.
* Collaborate on https://github.com/Bioconductor/GoogleGenomics.
